Add proposal deposits to DRep active voting stake.

aniketd opened the pull request:
# Description

Add proposal-deposits to the DRep distribution.

Partially addresses #4061. The same will be implemented for SPOs in a follow-up.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

aniketd committed:
Add proposal deposits to DRep active voting stake.

Also add imptests.

teodanciu commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -624,6 +627,117 @@ votingSpec =
+                    { dvtCommitteeNormal = 51 %! 100
+                    , dvtCommitteeNoConfidence = 51 %! 100
+                    }
+                & ppCommitteeMaxTermLengthL .~ EpochInterval 20
```
There is no need to modify `ppCommitteeMaxTermLengthL`, because it is set inthis way when initializing Conway.
I assume the `51 %! 100` threshold is relevant for the test, right?  But is the `dvtCommitteeNoConfidence` one relevant as well?

aniketd replied:
Thanks for pointing this out :+1:

teodanciu commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -624,6 +627,117 @@ votingSpec =
+            -- Submit another proposal to bump up the active voting stake
+            anotherCC <- KeyHashObj <$> freshKeyHash
+            let anotherAddCCAction = UpdateCommittee SNothing mempty (Map.singleton anotherCC 10) (75 %! 100)
+            _anotherAddCCGaid <-
```
Why not submitProposal_ instead of binding to this underscore name?

teodanciu commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -624,6 +627,117 @@ votingSpec =
+            passNEpochs 2
+            -- The vote should not result in a ratification
+            isDRepAccepted addCCGaid `shouldReturn` False
+            getLastEnactedCommittee `shouldReturn` SNothing
```
This is just a question for my education, that I was meaning to ask, because I came across it when doing some refactorings on the previous RP. 
I don't understand the meaning of these `getLastEnacted`* functions .
I can see they are looking in :
`nesEsL . esLStateL . lsUTxOStateL . utxosGovStateL . proposalsGovStateL . pRootsL`, and then `grCommitteeL . prRootL` for committee or `grConstitutionL . prRootL` for constitution, etc. 
So this would be the thing that *will* be enacted after one epoch?  What is the difference between this and looking in 
` nesEsL . esLStateL . lsUTxOStateL . utxosGovStateL . committeeGovStateL`  ?  Is the former a sort of preview of the latter?

aniketd replied:
Great questions!

> So this would be the thing that will be enacted after one epoch?

No. These are the last enacted governance action ids, as the name suggests. Which means, they are each returning the last governance action id that was enacted in the last epoch boundary.

>  looking in `nesEsL . esLStateL . lsUTxOStateL . utxosGovStateL . committeeGovStateL`?

This will give you the current state of the committee, as a whole, but not tell you anything about what was the last governance action that was enacted at the previous epoch boundary.

teodanciu replied:
Oh I see, so if nothing else was enacted in the meantime, I guess those two would coincide, but it's more precise to check with `getLastEnacted`, to verify that the action had indeed the desired effect (rather than it being set in the gov state for other reason). 
Thanks for the explanation!

aniketd replied:
> I guess those two would coincide

The functions `getLastEnacted...` return a `GovActionId`, whereas the `Committee` is just the committee. That is also the difference. So they are not the same thing.

teodanciu commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -624,6 +627,117 @@ votingSpec =
+                    { dvtCommitteeNormal = 51 %! 100
+                    , dvtCommitteeNoConfidence = 51 %! 100
+                    }
+                & ppCommitteeMaxTermLengthL .~ EpochInterval 20
```
Same comment about unnecessary setting of `ppCommitteeMaxTermLengthL` (and probably also of `dvtCommitteeNoConfidence`)

teodanciu commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -624,6 +627,117 @@ votingSpec =
+                  }
+            anotherCC <- KeyHashObj <$> freshKeyHash
+            let anotherAddCCAction = UpdateCommittee SNothing mempty (Map.singleton anotherCC 10) (75 %! 100)
+            _anotherAddCCGaid <-
```
Same as above, you could have less noise with submitProposal_

teodanciu commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -624,6 +627,117 @@ votingSpec =
+                  , pProcAnchor = def
+                  }
+            passNEpochs 2
+            -- The same vote should now successfully ratify the proposal
```
So is it the case that more money from the root utxo is now associated with this drep (via the deposit for this proposal), hence the vote of the drep has become more significant because of more stake?

aniketd replied:
Precisely, yes.

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -173,25 +174,42 @@ instance EraPParams era => FromCBOR (PulsingSnapshot era) where
   where
+    -- QUESTION: Can I memoize this as a field in DRepPulsar? That causes a
+    -- type-error, though that does _not_ get resolved even with AllowAmbiguousTypes
+    -- as `EraCrypto era` is a non-injective type-family
```
So you mean, in DRepPulser, instead of having
```dpProposals :: !(StrictSeq (GovActionState era))`, to have a map from credential to deposit?

aniketd replied:
Not "instead", but "in addition to", because we need `dpProposals` in an ordered container, as they are the signal to the `RATIFY` rule.

lehins replied:
The answer is it is an absolutele must to compute this during the DRep pulser setup in: `setFreshDRepPulsingState`. Feel free to ping me with a type error if you can't figure it out. 

You might need a `forall` in the `setFreshDRepPulsingState` and a type annotation on the field, but `AllowAmbiguousTypes` is definitely not the right solution here

aniketd replied:
This is now implemented, but I had to use `Proxy` :+1:

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -298,20 +316,31 @@ class
+        dpProposals
+        dpDRepDistr
+        $ Map.drop dpIndex
+        $ umElems dpUMap
```
A local binding will make it possible to avoid this column style
```suggestion
    !leftOver = Map.drop dpIndex $ umElems dpUMap
    !finalDRepDistr =
      computeDRepDistr dpStakeDistr dpDRepState dpProposals dpDRepDistr leftOver
```

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -298,20 +316,31 @@ class
             )
             globals
      in case ratifyResult of
           Left ps ->
```
```suggestion
          Left (x :| _) -> absurd x
```
then you can also get rid of the `error`

aniketd replied:
I did not fully understand the intention and motivation @Soupstraw, sorry :disappointed: . Could you explain a bit?

lehins replied:
`absurd` ensures at compile time that it is an impossible case, so you don't need to provide any error messages. This is possible because:
```haskell
type PredicateFailure (ConwayRATIFY era) = Void
```

aniketd committed:
Address review comments, add dpProposalDeposits.

field to the DRepPulser to memoize a `Map` from staking credentials to
total proposal deposits made by that credential.

aniketd committed:
Update changelog

aniketd committed:
Fix type signature of computeDRepDistr

aniketd committed:
Add changelog entries

aniketd committed:
Replace the call to error with absurd in RATIFY

lehins reviewed (approved):
Looks great. Thank you!
